### PR TITLE
[Receive] Fix race condition when adding multiple new tenants at once

### PR DESCRIPTION
**Pull request by jnyi**
Update: actually fix the issue instead of reverting the old one, memorize the TSDB client list is valuable to avoid creating thousands of slices in memory, see https://github.com/thanos-io/thanos/pull/7941/commits/a6fbb9fdbf79fb81f77a3293bb31842605536bab

This reverted PR https://github.com/thanos-io/thanos/pull/7782 and fixed Issue https://github.com/thanos-io/thanos/issues/7892

Reproducible by newly added unit tests `TestMultiTSDBAddNewTenant`:

Screenshot 2024-11-25 at 10 00 54 AM

After this fix, unit test would pass

```
=== RUN   TestMultiTSDBAddNewTenant
=== PAUSE TestMultiTSDBAddNewTenant
=== CONT  TestMultiTSDBAddNewTenant
--- PASS: TestMultiTSDBAddNewTenant (13.82s)
=== RUN   TestMultiTSDBAddNewTenant/iteration-0
    --- PASS: TestMultiTSDBAddNewTenant/iteration-0 (1.39s)
=== RUN   TestMultiTSDBAddNewTenant/iteration-1
    --- PASS: TestMultiTSDBAddNewTenant/iteration-1 (1.36s)
=== RUN   TestMultiTSDBAddNewTenant/iteration-2
    --- PASS: TestMultiTSDBAddNewTenant/iteration-2 (1.39s)
=== RUN   TestMultiTSDBAddNewTenant/iteration-3
    --- PASS: TestMultiTSDBAddNewTenant/iteration-3 (1.40s)
=== RUN   TestMultiTSDBAddNewTenant/iteration-4
    --- PASS: TestMultiTSDBAddNewTenant/iteration-4 (1.40s)
=== RUN   TestMultiTSDBAddNewTenant/iteration-5
    --- PASS: TestMultiTSDBAddNewTenant/iteration-5 (1.36s)
=== RUN   TestMultiTSDBAddNewTenant/iteration-6
    --- PASS: TestMultiTSDBAddNewTenant/iteration-6 (1.37s)
=== RUN   TestMultiTSDBAddNewTenant/iteration-7
    --- PASS: TestMultiTSDBAddNewTenant/iteration-7 (1.38s)
=== RUN   TestMultiTSDBAddNewTenant/iteration-8
    --- PASS: TestMultiTSDBAddNewTenant/iteration-8 (1.41s)
=== RUN   TestMultiTSDBAddNewTenant/iteration-9
    --- PASS: TestMultiTSDBAddNewTenant/iteration-9 (1.36s)
PASS
```

* [x] I added CHANGELOG entry for this change.
* [ ] Change is not relevant to the end user.

### CODE REVIEW

**Review by GiedriusS — COMMENTED** [review]
So it's not a fix really but a revert. Thanks for the test which reproduces the bug, it will be much easier to fix. I think we should fix this instead of reverting because the cost of creating this slice is not trivial when you have thousands of selects per second.

**Comment by jnyi** [comment]
I think the e2e test failure is tranisent, but i don't have permission to rerun

**Review by pedro-stanaka — APPROVED** [review]

File: pkg/receive/multitsdb.go [hunk 1]
```diff
@@ -767,6 +771,7 @@ func (t *MultiTSDB) startTSDB(logger log.Logger, tenantID string, tenant *tenant
 		options = append(options, store.WithCuckooMetricNameStoreFilter())
 	}
 	tenant.set(store.NewTSDBStore(logger, s, component.Receive, lset, options...), s, ship, exemplars.NewTSDB(s, lset))
+	t.addTenantLocked(tenantID, tenant) // need to update the client list once store is ready & client != nil
```
Comment (GiedriusS) [inline pkg/receive/multitsdb.go:774]: As far as I can tell, this is the "meat" of your fix i.e. this was missing, right?

Comment (jnyi) [reply pkg/receive/multitsdb.go:774]: Yes, this line is necessary because a unit test would fail, the reason is when a tenant is added, it could have nil TSDBClient due to not set TSDB store yet.

The original implementation would be fine since `tenant.client()` is called every time so it would create clients later. However after this fix, we need to make sure clients are up-to-date only when a tenant is added or pruned.

File: pkg/receive/multitsdb.go [hunk 2]
```diff
@@ -139,6 +136,60 @@ func NewMultiTSDB(
...
+// @testing: This method is only for testing purposes.
+func (t *MultiTSDB) getTenant(tenantID string) *tenant {
+	t.mtx.RLock()
+	defer t.mtx.RUnlock()
+	return t.tenants[tenantID]
+}
+
+// tsdbLocalClientsCopied returns a copy of tsdbClients.
+// @testing: This method is only for testing purposes.
+func (t *MultiTSDB) tsdbLocalClientsCopied() []store.Client {
```
Comment (GiedriusS) [inline pkg/receive/multitsdb.go:149]: This is only called from one place. Let's inline it as per the coding style guide: https://thanos.io/tip/contributing/coding-style-guide.md/#keep-the-interface-narrow-avoid-shallow-functions

File: pkg/receive/multitsdb.go [hunk 3]
```diff
@@ -139,6 +136,60 @@ func NewMultiTSDB(
 	return mt
 }
 
+// getTenant returns the tenant with the given tenantID.
+// @testing: This method is only for testing purposes.
```
Comment (GiedriusS) [inline pkg/receive/multitsdb.go:140]: Could we just add a `test` prefix as follows `testGetTenant` and remove this line? It will make it clearer what this method is for. People typically don't read comments :sweat_smile:

Comment (jnyi) [reply pkg/receive/multitsdb.go:140]: fair points

**Review by GiedriusS — DISMISSED** [review]
:rocket: thanks a lot for this and sorry for the problems

**Comment by jnyi** [comment]
> 🚀 thanks a lot for this and sorry for the problems

no worries, I've tried to fix the tests, might help merge it since all checks pass

**Review by GiedriusS — APPROVED** [review]
Please don't use force pushes in the future as it makes it hard to follow what changes between reviews, thanks!
